src/System.Private.ServiceModel/tests/Scenarios/Security/TransportSecurity/Tcp/ClientCredentialTypeCertificateCanonicalNameTests.cs

```diff
@@ -0,0 +1,119 @@
+            // DNS identity of the remote endpoint was 'localhost' but the remote endpoint provided DNS claim 'example.com'.If this
+            // is a legitimate remote endpoint, you can fix the problem by explicitly specifying DNS identity 'example.com' as
+            // the Identity property of EndpointAddress when creating channel proxy.
+
+            var errorBuilder = new StringBuilder();
+            errorBuilder.AppendFormat("The call to '{0}' should have been successful but failed with a MessageSecurityException. ", endpointUri.Host);
+            errorBuilder.AppendFormat("This usually means that the certificate validation failed when it should have passed. ");
+            errorBuilder.AppendFormat("When connecting to host '{0}', the expectation is that the DNSClaim will be for the same hostname. ", endpointUri.Host);
+            errorBuilder.AppendFormat("Exception message: {0}{1}", Environment.NewLine, exception.Message);
+
+            Assert.True(!shouldCallSucceed, errorBuilder.ToString());
+        }
+        finally
+        {
+            // *** ENSURE CLEANUP *** \\
+            ScenarioTestHelpers.CloseCommunicationObjects((ICommunicationObject)serviceProxy, factory);
+        }
+    }
+
+    public static List<object[]> TcpClientCredentialType_Certificate_With_Only_CanonicalName_MemberData
+    {
+        get
+        {
+            List<object[]> list = new List<object[]>();
+
+            var localhostEndpointUri = new Uri(Endpoints.Tcp_ClientCredentialType_Certificate_With_CanonicalName_Localhost_Address);
+            list.Add(new object[] { localhostEndpointUri, string.Compare(localhostEndpointUri.Host, "localhost", StringComparison.OrdinalIgnoreCase) == 0 });
+
+            var domainNameEndpointUri = new Uri(Endpoints.Tcp_ClientCredentialType_Certificate_With_CanonicalName_DomainName_Address);
+            list.Add(new object[] { domainNameEndpointUri, domainNameEndpointUri.Host.IndexOf('.') == -1 });
+
+            // Get just the hostname part of the domainName Uri
+            var domainNameHost = domainNameEndpointUri.Host.Split('.')[0];
+
+            // If the Bridge's reported FQDN is the same as the Bridge's reported hostname, 
+            // it means that there the bridge is set up on a network where FQDNs aren't used, only hostnames.
+            // Since our pass/fail detection logic on whether or not this is an FQDN depends on whether the host name has a '.', we don't test this case
+
+            var fqdnEndpointUri = new Uri(Endpoints.Tcp_ClientCredentialType_Certificate_With_CanonicalName_Fqdn_Address);
+            if (string.Compare(domainNameHost, fqdnEndpointUri.Host, StringComparison.OrdinalIgnoreCase) != 0)
+            {
+                list.Add(new object[] { fqdnEndpointUri, fqdnEndpointUri.Host.IndexOf('.') > -1 });
+            }
+
+            return list;
+        }
+    }
+}
```

Comment:
Deterministic in the run time is sufficient. Thanks for being creative to solve the issue, Jason! LGTM from testing perspective.